src/borrower/nftfeed/nftfeed.sol

```diff
@@ -0,0 +1,126 @@
+    function shelf(uint loan) public view returns (address registry, uint tokenId);
+}
+
+contract PileLike {
+    function setRate(uint loan, uint rate) public;
+    function debt(uint loan) public returns (uint);
+}
+
+contract NFTFeed is DSNote, Auth, Math {
+    // nftID => nftValues
+    mapping (bytes32 => uint) public nftValues;
+
+    // nftID => risk
+    mapping (bytes32 => uint) public risk;
+
+    // risk => thresholdRatio
+    mapping (uint => uint) public thresholdRatio;
+    // risk => ceilingRatio
+    mapping (uint => uint) public ceilingRatio;
+    // risk => rate
+    mapping (uint => uint) public rate;
+
+    PileLike pile;
+    ShelfLike shelf;
+
+    /// defines default values for risk group 0
+    /// all values are denominated in RAY (10^27)
+    constructor (uint defaultThresholdRatio, uint defaultCeilingRatio, uint defaultRate) public {
+        thresholdRatio[0] = defaultThresholdRatio;
+        ceilingRatio[0] = defaultCeilingRatio;
+        rate[0] = defaultRate;
+    }
+
+
+    /// sets the dependency to another contract
+    function depend(bytes32 contractName, address addr) external auth {
+        if (contractName == "pile") {pile = PileLike(addr);}
+        else if (contractName == "shelf") { shelf = ShelfLike(addr); }
+        else revert();
+    }
+
+    function nftID(address registry, uint tokenId) public pure returns (bytes32) {
+        return keccak256(abi.encodePacked(registry, tokenId));
+    }
+
+    function nftID(uint loan) public view returns (bytes32) {
+        (address registry, uint tokenId) = shelf.shelf(loan);
+        return nftID(registry, tokenId);
+    }
+
+    /// Admin -- Updates
+    function setRiskGroup(uint risk_, uint thresholdRatio_, uint ceilingRatio_, uint rate_) public auth {
+        thresholdRatio[risk_] = thresholdRatio_;
+        ceilingRatio[risk_] = ceilingRatio_;
+        rate[risk_]= rate_;
+    }
+
+    function file(bytes32 what, uint risk_, uint value_) public auth {
+        if (what == "threshold") { thresholdRatio[risk_]  = value_; }
+        else if (what == "ceiling") { ceilingRatio[risk_] = value_; }
+        else if (what == "rate") { rate[risk_]= value_;}
+        else revert("unknown parameter");
+    }
+
+    ///  -- Oracle Updates --
+
+    /// update the nft value
+    function update(bytes32 nftID_,  uint value) public auth {
+        nftValues[nftID_] = value;
+    }
+
+    /// update the nft value and change the risk group
+    function update(bytes32 nftID_, uint value, uint risk_) public auth {
+        require(thresholdRatio[risk_] != 0, "threshold for risk group not defined");
+        require(ceilingRatio[risk_] != 0, "ceiling for risk group not defined");
+        require(rate[risk_] != 0, "rate for risk group not defined");
+
+        risk[nftID_] = risk_;
+        nftValues[nftID_] = value;
+    }
+
+    // sets the loan rate in pile
+    // not possible for ongoing loans
+    function setPileRate(uint loan) external auth {
```

Comment:
We can do it like this or refactor the code and the pile gets the rate from the NFTFeed. (more elegant but more work)

Reply:
We need to be aware if we change the rate of a `risk group`  all ongoing loans of the risk group would still have the old rate in pile. Maybe that's okay, otherwise, we would have to call `changeRate` on all loans of the group.

If the pile would directly call the nftFeed for getting the rate. It would also be tricky if the rate changes (because it would require a recalculation of the `pie` (see change rate).

We could also say it is not possible to change the rate of a risk group instead move all members to a new risk group


We